File: benchmarking/tests/passport_processing_02/solution/solution.js
const fs = require('fs');

function countValidPassports(f) {
  const file = fs.readFileSync(f, "utf-8");
  const passes = file.split("\n\n").map(p => p.trim());
  return passes.filter(pass => isValidPassport(pass.split("\n").join(" "))).length;
}

function isValidPassport(pass) {
  return pass.split(" ")
        .map(f => f.split(":"))
        .filter(f => f[0] !== "cid")
        .filter(f => isValidField(f[0], f[1]))
        .length === 7;
}

function isValidField(field, value) {
    if (field === "byr") return +value >= 1920 && +value <= 2002;
    if (field === "iyr") return +value >= 2010 && +value <= 2020;
    if (field === "eyr") return value.length === 4 && +value >= 2020 && +value <= 2030;
    if (field === "hcl") return null !== value.match(/^#[0-9a-f]{6}$/);
    if (field === "ecl") return ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"].includes(value);
    if (field === "pid") return null !== value.match(/^[0-9]{9}$/);
    if (field === "hgt") {
      const v = value.substr(0, value.length-2);
      if (value.match(/cm$/)) {
        return +v >= 150 && +v <= 193;
      } else if (value.match(/in$/)) {
        return +v >= 59 && +v <= 76;
      }
      return false;
    };
}

File: benchmarking/tests/passport_processing_02/solution/solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

public class Submission {

    public static boolean isValidField(String key, String value) {
        switch (key){
            case "byr": {
                if(value.length() != 4) return false;
                int v = Integer.parseInt(value);
                return 1920 <= v && v <= 2002;
            }
            case "iyr": {
                if(value.length() != 4) return false;
                int v = Integer.parseInt(value);
                return 2010 <= v && v <= 2020;
            }
            case "eyr": {
                if(value.length() != 4) return false;
                int v = Integer.parseInt(value);
                return 2020 <= v && v <= 2030;
            }
            case "hgt": {
                if(value.contains("cm")) {
                    int v = Integer.parseInt(value.split("cm")[0]);
                    return 150 <= v && v <= 193;
                } else if(value.contains("in")) {
                    int v = Integer.parseInt(value.split("in")[0]);
                    return 59 <= v && v <= 76;
                }
                return false;
            }
            case "hcl": {
                return value.matches("#([0-9a-f]){6}");
            }
            case "ecl": {
                return Set.of("amb", "blu", "brn", "gry", "grn", "hzl", "oth").contains(value);
            }
            case "pid": {
                return value.matches("[0-9]{9}");
            }
            default:
                return true;
        }
    }


    public static boolean isValidPassport(String passport) {
        String[] pairs = passport.split(" ");
        Set<String> fields = new HashSet<String>();
        for (int i = 0; i < pairs.length; i++) {
            String[] pair = pairs[i].split(":");
            String key = pair[0];
            String value = pair[1];
            if(!isValidField(key, value)) return false;
            fields.add(key);
        }
        return fields.containsAll(Set.of("byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"));
    }


    public static int countValidPassports(String passports) {
        Path file = Path.of(passports);
        try {
            List<String> lines = Files.readAllLines(file);
            int count = 0;
            String passport = "";
            for (int i = 0; i < lines.size(); i++) {
                if(lines.get(i).isBlank()) {
                    if(isValidPassport(passport.strip())) count++;
                    passport = "";
                } else {
                    passport += " " + lines.get(i);
                }
            }
            if(!lines.get(lines.size()-1).isBlank()) {
                if(isValidPassport(passport.strip())) count++;
            }
            return count;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return -1;
    }
}
